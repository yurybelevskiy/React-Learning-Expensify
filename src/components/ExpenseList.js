import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (
                <p>No expenses</p>
            ) : (
                props.expenses.map((expense) => {
                    return (
                        <p key={expense.id}>
                            <ExpenseListItem {...expense} />
                        </p>
                    );
                })
            )
        }
    </div>
);

const mapStateToProps = (state) => {
    const expenses = selectExpenses(state.expenses, state.filters);
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList);