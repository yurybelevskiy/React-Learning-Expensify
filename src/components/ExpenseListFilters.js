import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilters extends React.Component {

    state = {
        calendarFocused: null
    };

    onPickerDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }

    onPickerFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    }

    onFilterTextChange = (e) => {
        this.props.setTextFilter(e.target.value)
    }

    onSortChange = (e) => {
        const sortBy = e.target.value;
        if (sortBy === "date") {
            this.props.sortByDate();
        } else if (sortBy === "amount") {
            this.props.sortByAmount();
        }
    }

    render() {
        return (
            <div>
                <input 
                    type="text"
                    value={this.props.filters.text} 
                    onChange={this.onFilterTextChange} 
                />
                <select
                    value={this.props.filters.sortBy}
                    onChange={this.onSortChange}
                >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker 
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onPickerDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onPickerFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    showClearDates={true}
                />
            </div>
        );
    }
};

const mapStateToProps = (state) => ({
    filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount())
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);