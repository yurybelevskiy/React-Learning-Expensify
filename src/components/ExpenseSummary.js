import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';

import getVisibleExpenses from '../selectors/expenses';
import calculateExpensesTotal from '../selectors/expenses-total';

export const ExpenseSummary = ({ expensesCount, expensesTotal }) => (
    <div>
        {
            expensesCount > 0 && 
            (
                <p>
                    {`
                        Viewing ${expensesCount} 
                        ${expensesCount === 1 ? "expense" : "expenses"} 
                        totalling ${numeral(expensesTotal / 100).format("$0,0.00")}
                    `}
                </p>
            )
        }
    </div>
);

const mapStateToProps = (state) => {
    const expenses = getVisibleExpenses(state.expenses, state.filters);
    return {
        expensesCount: expenses.length,
        expensesTotal: calculateExpensesTotal(expenses)
    };
};

export default connect(mapStateToProps)(ExpenseSummary);

