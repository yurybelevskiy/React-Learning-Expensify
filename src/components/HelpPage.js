import React from 'react';

const HelpPage = () => (
    <div>
        This is Help component
    </div>
);

export default HelpPage;