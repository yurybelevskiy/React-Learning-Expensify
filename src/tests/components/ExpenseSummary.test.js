import React from 'react';
import { shallow } from 'enzyme';

import { ExpenseSummary } from '../../components/ExpenseSummary';
import expenses from '../fixtures/expenses';

test("should render correctly when no expenses are added", () => {
    const wrapper = shallow(<ExpenseSummary expensesCount={0} expensesTotal={0} />);

    expect(wrapper).toMatchSnapshot();
});

test("should render correctly when single expense is added", () => {
    const wrapper = shallow(<ExpenseSummary expensesCount={1} expensesTotal={25} />);

    expect(wrapper).toMatchSnapshot();
});

test("should render correctly when multiple expenses are added", () => {
    const wrapper = shallow(<ExpenseSummary expensesCount={5} expensesTotal={5000} />);

    expect(wrapper).toMatchSnapshot();
});