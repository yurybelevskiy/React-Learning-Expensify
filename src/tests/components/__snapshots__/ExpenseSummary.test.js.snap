// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly when multiple expenses are added 1`] = `
<div>
  <p>
    
                        Viewing 5 
                        expenses 
                        totalling $50.00
                    
  </p>
</div>
`;

exports[`should render correctly when no expenses are added 1`] = `<div />`;

exports[`should render correctly when single expense is added 1`] = `
<div>
  <p>
    
                        Viewing 1 
                        expense 
                        totalling $0.25
                    
  </p>
</div>
`;
